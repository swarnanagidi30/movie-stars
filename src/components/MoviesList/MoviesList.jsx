import React from 'react'
import { makeStyles } from '@material-ui/core/styles';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import CircularProgress from '@material-ui/core/CircularProgress';
import Grid from '@material-ui/core/Grid';
import fetch from 'node-fetch';

const useStyles = makeStyles((theme) => ({
    root: {
        width: '100%',
        maxWidth: 360,
        minHeight: '30vh',
        marginLeft: 5
        // backgroundColor: theme.palette.background.paper,
    },
    title: {
        width: '100%',
        textAlign: 'left'
    }
}));

const allfilms = {};

const getFilmId = (film) => {
    if (!film) {
        return -1;
    }
    return film.replace('http://swapi.dev/api/films/', '').replace('/', '');
}

export const getFilmDetail = (films) => {
    return Promise.all(films.map((film) => {
        const id = getFilmId(film);
        return allfilms[id] ? Promise.resolve(allfilms[id]) : fetch(film.replace('http://','https://')).then(async (resp) => {
            const filmDetail = await resp.json();
            allfilms[getFilmId(filmDetail.url)] = filmDetail;
            return filmDetail;
        });
    }))
};


export default function MoviesList({ films, onLatestMovie }) {
    const classes = useStyles();
    const [filmDetails, setFilmDetails] = React.useState([]);
    React.useEffect(() => {
        if (films && films.length > 0) {
            getFilmDetail(films).then(f => {
                setFilmDetails(f);
            });
        }
    }, [films]);

    if (films && filmDetails.length === 0) {
        return <CircularProgress color="inherit" size={20} />;
    }
    if (onLatestMovie && filmDetails.length > 0) {
        setTimeout(() => onLatestMovie(filmDetails[filmDetails.length - 1]), 1000)
    }
    return (
        <Grid container spacing={2} className={classes.root}>
            <p className={classes.title}>
                List of Movies:
          </p>
            <List dense aria-label="contacts">
                {filmDetails && filmDetails.map((flim, index) => (
                    <ListItem key={index}>
                        <ListItemText inset primary={flim.title} />
                    </ListItem>
                ))}
            </List>
        </Grid>
    )
}
