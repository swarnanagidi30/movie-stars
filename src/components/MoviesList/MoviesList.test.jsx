import React from 'react';
jest.mock('node-fetch');

// import fetch, { Response } from 'node-fetch';

import { createShallow } from '@material-ui/core/test-utils';
import { ThemeProvider } from '@material-ui/core/styles';
import MoviesList from './MoviesList';
import toJson from "enzyme-to-json"

describe("<MoviesList />", () => {
    const renderShallowUntilComponent = createShallow({ "untilSelector": "MoviesList" })
    let Component
    beforeEach(() => {

        Component = renderShallowUntilComponent(<ThemeProvider theme={{ success: { main: '#fff' } }}><MoviesList /></ThemeProvider>)
    })

    it("loads", () => {
        expect(typeof Component).toBe("object")
    })

    it("has a <Grid> container", () => {
        expect(Component.find("WithStyles(ForwardRef(Grid))[container=true]").length).toBeGreaterThanOrEqual(1)
    })


    it("has a <List> container", () => {
        expect(Component.find("WithStyles(ForwardRef(List))").length).toBeGreaterThanOrEqual(1)
    })

    it("snapshots", () => {
        expect(toJson(Component)).toMatchSnapshot()
    })
})
