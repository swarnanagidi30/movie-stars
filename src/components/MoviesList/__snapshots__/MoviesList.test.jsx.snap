// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MoviesList /> snapshots 1`] = `
<WithStyles(ForwardRef(Grid))
  className="makeStyles-root-1"
  container={true}
  spacing={2}
>
  <p
    className="makeStyles-title-2"
  >
    List of Movies:
  </p>
  <WithStyles(ForwardRef(List))
    aria-label="contacts"
    dense={true}
  />
</WithStyles(ForwardRef(Grid))>
`;
